enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -2822,6 +2825,442 @@ class AdjointGenerator
+          auto mask = ConstantAggregateZero::get(VectorType::get(
+              Type::getInt32Ty(und->getContext()),
+#if LLVM_VERSION_MAJOR >= 11
+              cast<VectorType>(und->getType())->getElementCount()));
+#else
+              cast<VectorType>(und->getType())->getNumElements()));
+#endif
+          auto accdiff = diffe(orig_ops[0], Builder2);
+          auto diff = Builder2.CreateShuffleVector(
+              Builder2.CreateInsertElement(und, accdiff, (uint64_t)0), und,
+              mask);
+          setDiffe(&I, diff, Builder2);
+        }
+        return;
+      }
+#endif
+
+      case Intrinsic::lifetime_start:
+      case Intrinsic::nvvm_sqrt_rn_d:
+      case Intrinsic::sqrt: {
+        Value *op = diffe(orig_ops[0], Builder2);
+
+        if (op && !gutils->isConstantValue(orig_ops[0])) {
+          SmallVector<Value *, 2> args = {
+              lookup(gutils->getNewFromOriginal(orig_ops[0]), Builder2)};
+          Type *tys[] = {orig_ops[0]->getType()};
+          Function *SqrtF;
+          if (ID == Intrinsic::sqrt)
+            SqrtF = Intrinsic::getDeclaration(M, ID, tys);
+          else
+            SqrtF = Intrinsic::getDeclaration(M, ID);
+
+          auto cal = cast<CallInst>(Builder2.CreateCall(SqrtF, args));
+          cal->setCallingConv(SqrtF->getCallingConv());
+          cal->setDebugLoc(gutils->getNewFromOriginal(I.getDebugLoc()));
+
+          Value *dif0 = Builder2.CreateFDiv(
+              Builder2.CreateFMul(ConstantFP::get(I.getType(), 0.5), op), cal);
+
+          Value *cmp = Builder2.CreateFCmpOEQ(
+              args[0], ConstantFP::get(orig_ops[0]->getType(), 0));
+          dif0 = Builder2.CreateSelect(
+              cmp, ConstantFP::get(orig_ops[0]->getType(), 0), dif0);
+
+          setDiffe(&I, dif0, Builder2);
+        }
+        return;
+      }
+
+      case Intrinsic::fabs: {
+        Value *op = diffe(orig_ops[0], Builder2);
+
+        if (op && !gutils->isConstantValue(orig_ops[0])) {
+          Value *cmp = Builder2.CreateFCmpOLT(
+              lookup(gutils->getNewFromOriginal(orig_ops[0]), Builder2),
+              ConstantFP::get(orig_ops[0]->getType(), 0));
+          Value *dif0 = Builder2.CreateFMul(
+              Builder2.CreateSelect(cmp,
+                                    ConstantFP::get(orig_ops[0]->getType(), -1),
+                                    ConstantFP::get(orig_ops[0]->getType(), 1)),
+              op);
+          setDiffe(&I, dif0, Builder2);
+        }
+        return;
+      }
+
+#if LLVM_VERSION_MAJOR < 10
+      case Intrinsic::x86_sse_max_ss:
+      case Intrinsic::x86_sse_max_ps:
+#endif
+      case Intrinsic::maxnum: {
+        auto op0 = gutils->getNewFromOriginal(orig_ops[0]);
+        auto op1 = gutils->getNewFromOriginal(orig_ops[1]);
+        Value *cmp = Builder2.CreateFCmpOLT(op0, op1);
+
+        if (!gutils->isConstantValue(orig_ops[0]) &&
+            !gutils->isConstantValue(orig_ops[1])) {
```

Comment:
I'd change this to have a variable like this `diffe0 = isConstantValue(op[0]) ? 0 : diffe(op[0])`. It would simplify the case work here.

Reply:
Sorry, to clarify the 0 in the above should be a constant, not a literal zero [meaning nullptr]

Reply:
I have written way too much swift code where any object can be represented a custom literal. 😄